### ISSUE TITLE
c++14 support

### ISSUE THREAD

**Issue by nevion** [steven-varga/h5cpp#56]
I can definitely understand the usage of C++17, but it's still a bit too new for my usages.

I'm not sure how many c++17 features h5cpp uses but one I am seeing is  if-constexpr variations.

I see some attempt at constexpr version handling with h5cpp__constexpr, but it seems there's several  instances of the if constexprs.  I wonder if this is all that stands in the way?

**Comment by steven-varga** [comment]
No. Although in some cases the conditional constexpr is handy to optimise away lines which otherwise would result in compile time error -- an optional argument not being present. 

It is more like I don't have the resource to maintain backward compatibility and taking this project to where I planned it to be: a general efficient, scalable persistence for C++. Shedding extra weight helps me to move forward faster. C++20 is out (H5CPP is moving towards pure template based reflection), C++17 is supported by major compilers, and by the time I hit the target C++14 may be outdated. 

Are you on a platform where C++17 compatible compiler is not available?

**Comment by nevion** [comment]
>Although in some cases the conditional constexpr is handy to optimise away lines which otherwise would result in compile time error -- an optional argument not being present.

So you don't see another SFINAE based indirection to not-to-horribly handle the current more clean version?

>Are you on a platform where C++17 compatible compiler is not available?

essentially yes - cuda currently restricts C++ support to C++14.  Who knows when cuda on 17 will be out, otherwise I'd probably be on 17 as well.  Unfortunately that's sufficient to be a show stopper, for me.  If you're looking to survey the typical compiler language support though, you might take a look at the Eigen ML for various pollings of users, I suspect the community there could be a proxy for your prospective userbase.

Definitely understand the want to go to template based reflection but I think you might want to consider some mixed modes until maybe 1-2 years after compilers support it in as part of the standard language option set - right now it is just a TS, correct?

Are there any other parts besides the if-constexpr's that you know are c++17 only?  I'm only trying to establish if I make changes to handle c++14 support, how wide a surface that really is

**Comment by steven-varga** [comment]
Originally I kept up the C++14, but as of now it is rather costly. Also there are bigger problems such as to investigate LLVM failure etc... [Did you try to compile C++17 in gcc then link it against clang++](https://stackoverflow.com/questions/3217513/are-llvm-gcc-and-clang-binary-compatible-with-gcc-particularly-mingw-gcc-on-w)?

> Definitely understand the want to go to template based reflection but I think you might want to consider some mixed modes until maybe 1-2 years after compilers support it in as part of the standard language option set - right now it is just a TS, correct?

AFAIK h5cpp-compiler does exactly that: compiler assisted  reflection: making reflection available today until template based standardised reflection for c++ becomes available.  When I designed the LLVM tool 2 years ago I projected 6-8 years of lifespan. Indeed there is no template based reflection fully implemented yet -- however the next H5CPP will rely more on template based feature detection and eventually will shift to  concepts.  

> Are there any other parts besides the if-constexpr's that you know are c++17 only?

yes, the solution is to complete the phase I am working on then back port it -- if possible. The new version will have full STL, sparse matrix [, ...] support.
